plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -0,0 +1,295 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.settings.SettingsException;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.repositories.azure.executors.PrivilegedExecutor;
+import org.elasticsearch.repositories.azure.executors.ReactorScheduledExecutorService;
+import org.elasticsearch.threadpool.ThreadPool;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.netty.resources.ConnectionProvider;
+
+import java.io.IOException;
+import java.net.URL;
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.repositories.azure.AzureRepositoryPlugin.NETTY_EVENT_LOOP_THREAD_POOL_NAME;
+import static org.elasticsearch.repositories.azure.AzureRepositoryPlugin.REPOSITORY_THREAD_POOL_NAME;
+
+class AzureClientProvider extends AbstractLifecycleComponent {
+    private static final TimeValue DEFAULT_CONNECTION_TIMEOUT = TimeValue.timeValueSeconds(30);
+    private static final TimeValue DEFAULT_MAX_CONNECTION_IDLE_TIME = TimeValue.timeValueSeconds(60);
+    private static final int DEFAULT_MAX_CONNECTIONS = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
+    private static final int DEFAULT_EVENT_LOOP_THREAD_COUNT = Math.min(Runtime.getRuntime().availableProcessors(), 8) * 2;
+
+    static final Setting<String> EVENT_LOOP_EXECUTOR = Setting.simpleString(
+        "repository.azure.http_client.event_loop_executor_name",
+        NETTY_EVENT_LOOP_THREAD_POOL_NAME,
+        Setting.Property.NodeScope);
+
+    static final Setting<Integer> EVENT_LOOP_THREAD_COUNT = Setting.intSetting(
+        "repository.azure.http_client.event_loop_executor_thread_count",
+        DEFAULT_EVENT_LOOP_THREAD_COUNT,
+        Setting.Property.NodeScope);
+
+    static final Setting<Integer> MAX_OPEN_CONNECTIONS = Setting.intSetting(
+        "repository.azure.http_client.max_open_connections",
+        DEFAULT_MAX_CONNECTIONS,
+        Setting.Property.NodeScope);
+
+    static final Setting<TimeValue> OPEN_CONNECTION_TIMEOUT = Setting.timeSetting(
+        "repository.azure.http_client.connection_timeout",
+        DEFAULT_CONNECTION_TIMEOUT,
+        Setting.Property.NodeScope);
+
+    static final Setting<TimeValue> MAX_IDLE_TIME = Setting.timeSetting(
+        "repository.azure.http_client.connection_max_idle_time",
+        DEFAULT_MAX_CONNECTION_IDLE_TIME,
+        Setting.Property.NodeScope);
+
+    static final Setting<String> REACTOR_SCHEDULER_EXECUTOR_NAME = Setting.simpleString(
+        "repository.azure.http_client.reactor_executor_name",
+        REPOSITORY_THREAD_POOL_NAME,
+        Setting.Property.NodeScope);
+
+    private final ThreadPool threadPool;
+    private final String reactorExecutorName;
+    private final EventLoopGroup eventLoopGroup;
```

Comment:
All this resources are shared by all the clients, so we only need to manage the lifecycle of this component and the clients can be stateless.